modules/stochastic_tools/include/samplers/ParallelMarkovChainMonteCarloBase.h

```diff
@@ -0,0 +1,109 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Sampler.h"
+#include "ReporterInterface.h"
+#include "LikelihoodInterface.h"
+
+/**
+ * A base class used to perform Parallel Markov Chain Monte Carlo (MCMC) sampling
+ */
+class ParallelMarkovChainMonteCarloBase : public Sampler,
+                                          public ReporterInterface,
+                                          public LikelihoodInterface
+{
+public:
+  static InputParameters validParams();
+
+  ParallelMarkovChainMonteCarloBase(const InputParameters & parameters);
+
+  /**
+   * Return the number of configuration parameters.
+   */
+  dof_id_type getNumberOfConfigParams() const { return _confg_values.size(); }
+
+  /**
+   * Return the number of parallel proposals.
+   */
+  dof_id_type getNumParallelProposals() const { return _num_parallel_proposals; }
+
+  /**
+   * Return the random numbers to facilitate decision making in reporters
+   */
+  const std::vector<Real> & getRandomNumbers() const;
+
+  /**
+   * Return the step after which decision making can begin
+   */
+  virtual int decisionStep() const { return 1; }
+
+protected:
+  // Fill in the _new_samples vector of vectors (happens within sampleSetUp)
+  virtual void proposeSamples(const unsigned int seed_value);
+
+  // See Sampler.h for description
+  virtual void sampleSetUp(const Sampler::SampleMode mode) override;
+
+  // See Sampler.h for description
+  virtual Real computeSample(dof_id_type row_index, dof_id_type col_index) override;
+
+  /// Sample a random index excluding a specified index
+  void randomIndex(const unsigned int & ub,
+                   const unsigned int & exclude,
+                   const unsigned int & seed,
+                   unsigned int & req_index);
```

Comment:
```suggestion
  unsigned int randomIndex(unsigned int ub, // <- what does "ub"mean?
                   unsigned int exclude_index,
                   unsigned int seed);
```

Reply:
It is an upper bound integer.

Reply:
what are you indexing into?

Reply:
Please use a more descriptive variable name like `upper_bound`